animations/shared-element-transitions/index.md

```diff
@@ -0,0 +1,684 @@
+the current route, obscuring it during flight.
+
+<aside class="alert alert-info" markdown="1">
+**Note about dialogs:**
+Heroes fly from one PageRoute to another. Dialogs
+(displayed with `showDialog()`, for example), use PopupRoutes,
+which are not PageRoutes.  At least for now,
+you can't animate a Hero to a Dialog.
+For further developments (and a possible workaround), [watch this
+issue.](https://github.com/flutter/flutter/issues/10667)
+</aside>
+
+Hero animation code has the following structure:
+
+1. Define a starting Hero object, referred to as the _source
+   Hero_. The Hero specifies its location,
+   size, graphical representation (typically an image), and an
+   identifying tag.
+1. Define an ending Hero object, referred to as the _destination Hero_.
+   This Hero also specifies its location, size, graphical
+   representation, and the same tag as the source Hero.
+   It's <strong>essential that both Hero objects are created with
+   the same tag</strong>, typically an object that represents the
+   underlying data. For best results, the Heroes should have
+   virtually identical widget trees.
+1. (Optional) Place any Hero object that responds to user input,
+   such as a tap, in an Inkwell.
+1. Specify a route that contains the destination Hero.
+   The destination route defines its widget tree at the end
+   of the transition.
+1. Trigger the animation by pushing the route on the Navigator's stack.
+   Popping a route from the Navigator’s stack triggers a return
+   transition.
+
+Flutter calculates the tweens required to transition the
+starting Hero to the ending Hero (interpolating size and
+position changes), and performs the animation in an overlay.
+
+The next section describes Flutter's process in greater detail.
+
+## Behind the scenes
+
+The source route contains the Hero inside an Inkwell.
+What happens when the InkWell receives a tap gesture?
+The following describes how Flutter performs the transition.
+
+<img src="images/hero-transition-0.png" alt="Before the transition the source hero appears in the source route">
+
+Before transition, the source Hero waits in the source route's widget
+tree. The destination route does not yet exist, and the overlay
+is blank and disabled.
+
+---
+
+<img src="images/hero-transition-1.png" alt="Transition begins and Destination Hero appears in the overlay">
+
+Tapping the Hero's InkWell triggers the animation, and the
+following occurs:
+
+* Offscreen, Flutter calculates the transition (including
+  position and size changes) from beginning to end,
+  moving the Hero along a curved path as described in the
+  Material motion spec. Flutter saves the size and location
+  of the Hero at the beginning and end of the transition.
+
+* Flutter places the destination Hero in the overlay, at the
+  same location and size as the _source_ Hero.
+
+* Flutter enables the overlay, obscuring the current route.
+
+* Flutter creates the destination route, placing it at the
```

Comment:
The destination Hero is part of the destination route's widget tree. I don't see how it can be available for tween calculations, if the destination route isn't created earlier than this.

Reply:
Right. At animation t=0.0 we create and layout the entire destination route "offscreen". This is done to figure out where the Hero is going to end up. Still at animation t=0.0 we move the destination Hero into the overlay and move the source hero offscreen. Then we start the hero animation. When the animation completes at t=1.0 we move both Heroes widgets back to the routes they came from (in other words at t=1.0 the destination hero moves from the overlay back to the destination route).